### ISSUE TITLE
Multiple DisconnectedContext exceptions after download

### ISSUE THREAD

**Issue by jannesen** [cefsharp/CefSharp#3203]
- **What version of the product are you using?**
    CefSharp.WinForms 84.3.30-CI3634 

- **What architecture x86 or x64?**
    x64
    
- **On what operating system?**
    Win10

- **Are you using `WinForms`, `WPF` or `OffScreen`?**
    WinForms
    
- **What steps will reproduce the problem?**
    CefSharp.MinimalExample.WinForm add a basis DownloadHandler (copy of CefSharp.MinimalExample.WinForm). Start de program in visual studio. Goto a page with a download link. For example https://github.com/cefsharp/CefSharp 'Download ZIP'. Download the zip and close the program. Or try to download the file multiple times at the 2e/3e time multiple Visual studio break with Managed Debugging Assistans.DisconnectedContext. Visual Studio is not able display the call stack.

- **What is the expected output? What do you see instead?**
No Managed Debugging Assistans DisconnectedContext exceptions.

- **Please provide any additional information below.**
The CefSharp.MinimalExample.WinForm with DownloadHandler I use
[CefSharp.MinimalExample.zip](https://github.com/cefsharp/CefSharp/files/5006999/CefSharp.MinimalExample.zip)

**Comment by amaitland** [comment]
The `Chromium Embedded Framework (CEF)` project has it's own [Issue Tracker](https://bitbucket.org/chromiumembedded/cef/issues), please report this issue there. `CefSharp` is just one of many wrappers around the [Chromium Embedded Framework (CEF)](https://bitbucket.org/chromiumembedded/cef/)

If you do find an existing issue that matches this one or create a new one then post a link here for others. 

Closing this as `upstream` as changes/fixes need to be made in `CEF`.

**Comment by jannesen** [comment]
Managed Debugging Assistans.DisconnectedContext is a exception from CLR! So it is CefSharp issue not a CEF issue.
I suspected that the callback parameters of OnBeforeDownload and OnDownloadUpdated are not properly releasing the underlining the COM object. Then when the thread that is handeling the Download request is terminated and de GC comes by DisconnectedContext is throw because the underlining COM object is no more.

See https://www.mirabulus.com/it/blog/2020/03/01/how-to-find-source-of-disconnected-context-with-managed-debugging-assistant.

**Comment by amaitland** [comment]
> Please no binary attachments (zip, 7z, etc), code needs to be easily reviewed in a web browser.

As per the [Bug Report Template](https://github.com/cefsharp/CefSharp/blob/master/.github/ISSUE_TEMPLATE/bug_report.md#bug-report) please no zip files. Please push your changes to GitHub. 

You will need to provide a stack trace also.

We don't actually use COM, the project uses VC++ to directly integrate with the CEF API.

**Comment by jannesen** [comment]
I have uploaded the test to https://github.com/jannesen/CefSharp.MinimalExample.

Visual Studio sometimes gives just 1 entry for stack trace:
> System.Management.dll!System.Management.ManagementObjectCollection.~ManagementObjectCollection() 
The active thread is GC Finalizer Thread.

I suspect that the cause is in the callback object CefBeforeDownloadCallbackWrapper and CefDownloadItemCallbackWrapper. These have a reference to CEF objects and that this object will not be released correctly once the download has finished. Then when de GC cleans up the DisconnectedContext  are thrown.

**Comment by amaitland** [comment]
Are you attempting to simultaneously download the files? Last time I checked CEF doesn't handle simultaneous downloads correctly. 

> I suspect that the cause is in the callback object CefBeforeDownloadCallbackWrapper and CefDownloadItemCallbackWrapper. These have a reference to CEF objects and that this object will not be released correctly once the download has finished.

You can explicitly dispose both when you are done, this is the general recommendation.

**Comment by amaitland** [comment]
I'd suggest you enable native debugging in Visual studio and download the libcef pdb 

https://github.com/cefsharp/CefSharp/wiki/Trouble-Shooting#loading-native-symbols-for-easier-diagnosis

It's common for the managed exception to be misleading.

**Comment by jannesen** [comment]
I found the cause! The culprit was 'WD Desktop App' which was installed as part of the installation of a 'WD My Book Duo'. A part of the WD Desktop App is a managed shell extension. This extension was loaded into the applicant when opening a OpenFileDialog (part of BeforeDownloadCallback.Continue).
This managed shell extension caused the DisconnectedContext events! So I uninstalled WD Desktop App and no more problems.

To find this by using WinDbg as described in https://www.mirabulus.com/it/blog/2020/03/01/how-to-find-source-of-disconnected-context-with-managed-debugging-assistant.
